homeassistant/components/dwd_weather_warnings/sensor.py

```diff
@@ -9,137 +9,174 @@
+    """Import the configurations from YAML to config flows."""
+    # Show issue as long as the YAML configuration exists.
+    async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        is_fixable=False,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+    )
+
+    # Only import the deprecated config, if it hasn't been done yet.
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if entry.source == SOURCE_IMPORT:
+            return
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config
+        )
+    )
+
 
-    api = WrappedDwDWWAPI(DwdWeatherWarningsAPI(region_name))
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up entities from config entry."""
+    name: str = entry.data[CONF_NAME]
+    monitored_conditions: list[str] = entry.data[CONF_MONITORED_CONDITIONS]
+    api = hass.data[DOMAIN][entry.entry_id]
+    coordinator = DwdWeatherWarningsCoordinator(hass, entry, api, name)
 
-    sensors = [
-        DwdWeatherWarningsSensor(api, name, description)
-        for description in SENSOR_TYPES
-        if description.key in config[CONF_MONITORED_CONDITIONS]
-    ]
+    # Add sensor for every monitored condition.
+    entities: list[DwdWeatherWarningsSensor] = []
+    for description in SENSOR_TYPES:
+        if description.key in monitored_conditions:
+            entities.append(DwdWeatherWarningsSensor(coordinator, description, name))
 
-    add_entities(sensors, True)
+    async_add_entities(entities, True)
 
 
-class DwdWeatherWarningsSensor(SensorEntity):
-    """Representation of a DWD-Weather-Warnings sensor."""
+class DwdWeatherWarningsCoordinator(DataUpdateCoordinator):
+    """Custom coordinator for the dwd_weather_warnings integration."""
 
     def __init__(
         self,
-        api,
-        name,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        api: DwdWeatherWarningsAPI,
+        name: str,
+    ) -> None:
+        """Initialize the dwd_weather_warnings coordinator."""
+        super().__init__(hass, LOGGER, name=name, update_interval=DEFAULT_SCAN_INTERVAL)
+
+        self.api = api
+        self.config_entry = entry
+
+    async def _async_update_data(self) -> dict:
+        """Get the latest data from the DWD Weather Warnings API."""
+        if self.config_entry is not None:
+            if self.config_entry.data.get(CONF_GPS_ACTIVE, False):
+                position = get_position_data(
+                    self.hass, self.config_entry.data[CONF_GPS_TRACKER]
+                )
+                self.api = await self.hass.async_add_executor_job(
+                    DwdWeatherWarningsAPI, position
+                )
+            else:
+                await self.hass.async_add_executor_job(self.api.update)
+
+        return self.api
```

Comment:
api is not a dict

Reply:
Can I just remove the return here or change it to an empty dict? I'm using the coordinators `api` variable directly in the entity class anyway. Otherwise I'd have to build my own dict with the data I guess which is not that clean imo

Reply:
Yup, you can return empty dict.